xlators/storage/posix/src/posix-entry-ops.c

```diff
@@ -953,7 +997,19 @@ posix_mkdir(call_frame_t *frame, xlator_t *this, loc_t *loc, mode_t mode,
         dict_del_sizen(xdata, GF_PREOP_PARENT_KEY);
     }
 
-    op_ret = sys_mkdir(real_path, mode);
+    if (!was_present && (frame->root->pid != GF_SERVER_PID_TRASH)) {
+        index = uuid_req[0];
+        uuid_str = uuid_utoa(uuid_req);
+        snprintf(gfid_tmp_path, sizeof(gfid_tmp_path), "%s/%s/%02x/tt/%s",
+                 priv->base_path, GF_HIDDEN_PATH, uuid_req[0], uuid_str);
+        snprintf(gfid_open_path, sizeof(gfid_open_path), "tt/%s", uuid_str);
+        op_ret = sys_mkdirat(priv->arrdfd[index], gfid_open_path, mode);
+        real_path = gfid_tmp_path;
+        atomic_creation = _gf_true;
+    } else {
+        op_ret = sys_mkdir(real_path, mode);
```

Comment:
Couldn't be a race here if the directory was present before, but removed just before calling `mkdir()` ?

Reply:
done